### ISSUE TITLE
Decide whether to keep SortedSetSubject and SortedMapSubject

### ISSUE THREAD

**Issue by cpovirk** [google/truth#556]
Each is used from only 2 classes internally. That's well below what I'd hope for for a method, let alone a whole class. Not that "a whole class" is that big a deal, but I feel at times bad about the number of methods in lists like https://google.github.io/truth/api/latest/index.html and https://google.github.io/truth/api/latest/com/google/common/truth/Truth.html, which make it hard to easily skim the available methods.

Somewhat related: We'd left open the possibility that we'd move these classes' methods up to the supertypes, like `IterableSubject.hasFirstElement`. Doing so would make them usable with non-`Sorted`-but-ordered collections, including `ImmutableSet`, list, etc. This is somewhat like AssertJ's [`first()`](http://joel-costigliola.github.io/assertj/core-8/api/org/assertj/core/api/AbstractIterableAssert.html#first--). I certainly see lots of hits for `assertThat[(].*[.]get[(]0[)][)]` (though some of them surely followed by an assertion on `get(1)` -- I see about 50% as many on `get(1)` as `get(0)`, though I haven't confirmed that they're adjacent) and some for `assertThat[(].*[.]iterator[(][)][.]next[(][)][)]`.

If nothing else, these are APIs that would be easy to omit for 1.0.

@dimo414

**Comment by cpovirk** [comment]
I feel somewhat similarly about `AtomicLongMapSubject`, but:

- It has more callers (though still very few by the standard of Truth assertions).
- Its methods couldn't be pulled up into a supertype.
- Users are probably less likely to look at it unless they already have an `AtomicLongMap`.
- I may be biased by feeling sort of bad about `AtomicLongMap` in general, which I've always felt guilty is so similar to `Multiset`.

So I would probably keep it.

@kluever

**Comment by cpovirk** [comment]
cross-link: https://github.com/google/truth/issues/249

**Comment by dimo414** [comment]
Perhaps we could start by trying to migrate the `assertThat\(.*\.get\(0\)\)` (I think that's equivalent to the `[]` notation) to use `Sorted*Subject`. Although I see the potential benefit of moving these methods up to the parent subject type, I worry they could lead to misuse (i.e. asserting on the ordering of `HashMap` etc.).

Other than the "number of methods" argument, is there a particular motivation for removing them?

**Comment by cpovirk** [comment]
`get(0)` works on `List` but not `SortedSet`/`SortedMap`, but we could still look at the `iterator().next()` hits. That said, I see a very small number of assertions on the result of `firstKey()`/`firstEntry()`/`first()` (and their `last*` equivalents), which `SortedSet`/`SortedMap` users would likely be using directly. I see that I'd done a search for them before when we were considering adding this, but it looks like I looked for _all_ calls to methods with those names in tests, even if they weren't the same methods, they weren't in assertions, or they were in assertions but the assertion was on the value of, say, `firstKey().foo()`.

"Number of methods" is basically my only argument. (OK, if we're keeping them, we'd ideally update them to the new key-value format, too. But I wouldn't remove them just because of that.) I just see it as an increasingly strong argument over time, particularly when I look at [the many other methods users are likely to request](http://joel-costigliola.github.io/assertj/core-8/api/org/assertj/core/api/AbstractIterableAssert.html), especially for collections. And again, I think I overestimated the number of callers this would have, thanks to my broad search.

I'm perfectly happy to keep the subjects around internally while we look into this more, but for purposes of release a 1.0, I'd want to back them out externally unless we can convince ourselves that there "should" be a lot more callers.
